spi/max7219_32x8_spi/max7219_32x8_spi.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2022 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "pico/binary_info.h"
+#include "hardware/spi.h"
+
+/* Example code to talk to a Max7219 driving an 32x8 pixel LED display via SPI
+
+   NOTE: The device is driven at 5v, but SPI communications are at 3v3
+
+   Connections on Raspberry Pi Pico board and a generic Max7219 board, other
+   boards may vary.
+
+   * GPIO 17 (pin 22) Chip select -> CS on max7219 board
+   * GPIO 18 (pin 24) SCK/spi0_sclk -> CLK on max7219 board
+   * GPIO 19 (pin 25) MOSI/spi0_tx -> DIN on max7219 board
+   * 5v (pin 36) -> VCC on max7219 board
+   * GND (pin 38)  -> GND on max7219 board
+
+   Note: SPI devices can have a number of different naming schemes for pins. See
+   the Wikipedia page at https://en.wikipedia.org/wiki/Serial_Peripheral_Interface
+   for variations.
+
+*/
+
+// This defines how many Max7219 modules we have cascaded together, in this case, we have 4 x 8x8 matrices giving a total of 32x8
+#define NUM_MODULES 4
+
+const uint8_t CMD_NOOP = 0;
+const uint8_t CMD_DIGIT0 = 1; // Goes up to 8, for each line
+const uint8_t CMD_DECODEMODE = 9;
+const uint8_t CMD_BRIGHTNESS = 10;
+const uint8_t CMD_SCANLIMIT = 11;
+const uint8_t CMD_SHUTDOWN = 12;
+const uint8_t CMD_DISPLAYTEST = 15;
+
+#ifdef PICO_DEFAULT_SPI_CSN_PIN
+static inline void cs_select() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 0);  // Active low
+    asm volatile("nop \n nop \n nop");
+}
+
+static inline void cs_deselect() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 1);
+    asm volatile("nop \n nop \n nop");
+}
+#endif
+
+#if defined(spi_default) && defined(PICO_DEFAULT_SPI_CSN_PIN)
+static void write_register(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+    sleep_ms(1);
+}
+
+static void write_register_all(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    for (int i = 0; i< NUM_MODULES;i++)
+        spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+}
+
+#endif
+
+int main() {
+    stdio_init_all();
+
+#if !defined(spi_default) || !defined(PICO_DEFAULT_SPI_SCK_PIN) || !defined(PICO_DEFAULT_SPI_TX_PIN) || !defined(PICO_DEFAULT_SPI_RX_PIN) || !defined(PICO_DEFAULT_SPI_CSN_PIN)
```

Comment:
`PICO_DEFAULT_SPI_RX_PIN` isn't used, so no need to check if it's defined?

Reply:
All this default stuff was added after I wrote various SPI example - not sure why they were added,. My preference is to remove all references to the default stuff here? Or do we always want to use defaults? Thoughts?

Reply:
Ah, fine to stick with defaults actually. All this stuff is boilerplate, same as other examples.

Reply:
Yeah, it allows people to e.g. modify `PICO_DEFAULT_SPI_SCK_PIN` in the CMakeLists.txt file for their particular board (or by using one of the [board header files](https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards)), so that they can use all the SPI examples without having to edit the source-code of the examples themselves.